source/_posts/2018-08-06-Android WorkManager.md

```diff
@@ -0,0 +1,132 @@
+</br>
+</br>Also a very important note about [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager): </br> *“WorkManager is intended for tasks that require a guarantee that the system will run them even if the app exits...”*
+
+
+## How to use it
+
+First you have to add the dependency in your `build.gradle` file: </br>
+`implementation "android.arch.work:work-runtime:1.0.0-alpha02"`
+
+We will look at the following:
+##### - How to create a Worker </br>- How to create Constraints</br>- How to schedule all Worker's to be executed in a particular order </br>- How to give a Worker instance an input and get its output
+
+### User story 
+
+To demonstrate how to use `WorkManager` we will start by establishing a made-up user story.
+In our imaginary app QuickSnapper, we press a shutter button and the app will take a picture and apply some stickers on it and upload it all in one automatic process thanks to the WorkManager.
+
+So let's split up the user story in 3 use cases:
+
+- 1 The user takes a photo (We want to compress it here) 
+- 2 The app adds weather and location information on the picture after step 1 (GPS and Internet must be available)
+- 3 The app uploads the image immediately after step 1-2  (Internet must be available) 
+
+
+For each use case we will make a `Worker` class. To do that we need to make a class and extend `Worker` which requires us to implement a `doWork()` method with a return type of a `WorkerResult` that can either be `WorkerResult.SUCCESS` or `WorkerResult.FAILURE`
+
+### Creating Workers
+
+#### 1) The first Worker compress our Bitmap into a smaller size, convert the Bitmap to ByteArray passes it in the WorkManager's outputData object and then we return WorkerResult.SUCCESS or WorkerResult.FAILURE.
+
+```kotlin
+    class ImageCompressionTask(val bitmap: Bitmap?) : Worker() {
+        override fun doWork(): WorkerResult {
+            val newBitmap: Bitmap?
+            try {
+                //Create a compressed bitmap
+                newBitmap = Bitmap.createScaledBitmap(bitmap, 500, 500, false)
+                return WorkerResult.SUCCESS
+            } catch (e: IllegalArgumentException) {
+                return WorkerResult.FAILURE
+            }
+        }
+    }
+```
+
+
+#### 2) In the second Worker we retrive the Bitmap from the Workers inputData object and adds some weather and location stickers on the image
+```kotlin
+    class AddStickersTask : Worker() {
```

Comment:
Can we add something about WorkerResult.RETRY? Maybe if it failed to get a location?